### PR TITLE
perf: gc refs for js tracing

### PR DESCRIPTION
**Pull request by mattsse**
ref #5241

we're currently taking a big perf Hit because we need to clone:
* memory
* state
* stack
per step, which is very expensive.

The reason for this is, the js engine requires ownership or `&'static` refs, because everything can be garbage collected, so capturing `&'a stack` in a javascript function would be unsafe.

This provides a solution by guarding access to a transmuted borrow that invalidates the reference if dropped resolving to a `None`.

While it would be a allowed during js tracing to capture the stack object for example:

```js
{
                stacks: [],
                step: function(log) {
                        this.stacks.push(log.stack);
                    }
                }
                result: function(ctx) {
                    // peek would now access a dropped gc ref which would result in an error
                    // this is also how geth behaves
                    return bigInt(this.stacks[0].peek(i)).toString(16);
                }
}

```

### CODE REVIEW

File: crates/revm/revm-inspectors/src/tracing/js/bindings.rs [hunk 1]
```diff
@@ -54,11 +54,77 @@ macro_rules! js_value_capture_getter {
...
+    }
+
+    /// Executes the given closure with a reference to the inner value if it is still present.
+    pub(crate) fn with_inner<F, R>(&self, f: F) -> Option<R>
+    where
+        F: FnOnce(&Val) -> R,
+    {
+        let inner = RefCell::borrow(&self.inner);
+        let inner = inner.as_ref()?;
+        Some(f(inner))
```
Comment (DaniPopes) [inline crates/revm/revm-inspectors/src/tracing/js/bindings.rs:99]:
```suggestion
        self.inner.borrow().map(f)
```

File: crates/revm/revm-inspectors/src/tracing/js/bindings.rs [hunk 2]
```diff
@@ -264,15 +330,39 @@ impl From<u8> for OpObj {
 }
 
 /// Represents the stack object
-#[derive(Debug, Clone)]
-pub(crate) struct StackObj(pub(crate) Stack);
+#[derive(Debug)]
+pub(crate) struct StackRef(pub(crate) GuardedNullableGcRef<Stack>);
```
Comment (DaniPopes) [inline crates/revm/revm-inspectors/src/tracing/js/bindings.rs:334]: Can we do the same to memory too?

**Review by DaniPopes — CHANGES_REQUESTED** [review]
Can we add tests for this and/or verify this is actually faster?

File: crates/revm/revm-inspectors/src/tracing/js/bindings.rs [hunk 3]
```diff
@@ -229,12 +238,15 @@ impl MemoryObj {
                     }
                     let start = start as usize;
                     let end = end as usize;
+                    let size = end - start;
```
Comment (DaniPopes) [inline crates/revm/revm-inspectors/src/tracing/js/bindings.rs:241]: maybe move conversions before condition

**Comment by mattsse** [comment]
tested on sepolia,
it's definitely faster, debug_traceBlockByNumber went down from ~3min to 11s.
this is still less performant than other clients, but it's a big improvement.
bottleneck now might just be the js engine

File: crates/revm/revm-inspectors/src/tracing/js/bindings.rs (hunk 3 above)
Comment (mattsse) [reply crates/revm/revm-inspectors/src/tracing/js/bindings.rs]: hmm, the condition checks for `< 0` since js numbers are signed.

**Comment by DaniPopes** [comment]
> tested on sepolia, it's definitely faster, debug_traceBlockByNumber went down from ~3min to 11s. this is still less performant than other clients, but it's a big improvement. bottleneck now might just be the js engine


Awesome, you can run [samply](https://github.com/mstange/samply) on it to see the hotspots

**Comment by mattsse** [comment]
I also think with the ref guard we could get rid of the js service task and instead capture the `&DB` instead, but can do this seprately.

File: crates/revm/revm-inspectors/src/tracing/js/bindings.rs (hunk 3 above)
Comment (DaniPopes) [reply crates/revm/revm-inspectors/src/tracing/js/bindings.rs:241]: ah it's fine, you also need it for formatting the error

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**